Handle parsing wildcard urls in launch profiles

davidfowl opened the pull request:
## Description

- As part of this change expose TargetHostAddress on EndpointAnnotation which describes the original address that was bound to in both the kestrel endpoint list and the launch profile's application urls.
- Added a test

Fixes #5587

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [x] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [x] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [x] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5588)

davidfowl committed:
Handle parsing wildcard urls in launch profiles
- As part of this change expose TargetHostAddress on EndpointAnnotation which describes the original address that was bound to in both the kestrel endpoint list and the launch profile's application urls.
- Added a test

Fixes #5587

davidfowl commented:
cc @davidebbo

davidfowl commented on src/Aspire.Hosting/ApplicationModel/EndpointAnnotation.cs:
```diff
@@ -104,6 +105,11 @@ public int? TargetPort
+    /// <summary>
+    /// This is the address the resource is listening on. By default it is localhost.
+    /// </summary>
+    public string TargetHostAddress { get; set; } = "localhost";
```
New public API. I could keep internal, but it felt better to expose this in WithEndpoint callbacks. This is to address as targetport is to port.

davidebbo replied:
Would `TargetAddressHost` be a more correct name? Or just `TargetHost`? If we go with the `BindingAddress` terminology, 'address' represents everything, and the host is just the one part that's the host name or IP address.

davidfowl replied:
I like TargetHost.

mitchdenny commented:
What impact does this have around requests that are proxied via DCP? I would have thought that for this to be useful the proxy would need to pass through a host header?

davidebbo commented on tests/Aspire.Hosting.Tests/ProjectResourceTests.cs:
```diff
@@ -590,6 +590,34 @@ public async Task AddProjectWithArgs()
+        var appModel = app.Services.GetRequiredService<DistributedApplicationModel>();
+        var projectResources = appModel.GetProjectResources();
+
+        var resource = Assert.Single(projectResources);
```
Should exercise new prop, e.g.

```c#
        foreach (var annotation in resource.Annotations.OfType<EndpointAnnotation>())
        {
            Assert.Equal("*", annotation.TargetHostAddress);
        }
```

davidfowl commented:
> What impact does this have around requests that are proxied via DCP? I would have thought that for this to be useful the proxy would need to pass through a host header?

Doesn't affect it. All we are doing is preserving the input. When you bind to * or 0.0.0.0, requests from localhost still work.

davidfowl committed:
PR feedback

davidfowl committed:
Tell DCP about the target host
- We still treat this allocated endpoint as being accessible via localhost only in the ux

davidfowl committed:
Fixed tests

karolz-ms commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -931,7 +931,7 @@ private void AddAllocatedEndpointInfo(IEnumerable<AppResource> resources)
                 sp.EndpointAnnotation.AllocatedEndpoint = new AllocatedEndpoint(
                     sp.EndpointAnnotation,
-                    sp.EndpointAnnotation.IsProxied ? svc.AllocatedAddress! : "localhost",
+                    "localhost",
```
The reason why `AllocatedAddress` is used here is because if one requests `localhost:1234`, but 1234 is already taken on one of the addresses behind `localhost` name (dual-stack machine), DCP will bind to specific IP address (IPv4 or IPv6) to satisfy the requested port. It will then return the specific IP address that was used.

If the client uses `localhost` to connect in that case, this will still work, but may be quite slow as the client tries the "wrong" address several times before switching to the other one.

I am not sure if you need to change this line as part of this change

davidfowl replied:
Will do

davidfowl replied:
This doesn't work because we use this allocated endpoint address to communicate with these endpoints. It seems like dcp returns 0.0.0.0 instead of localhost, 127.0.0.1 or ::1.

karolz-ms replied:
Returns 0.0.0.0 for what kind of data in EndpointAnnotation?

davidfowl replied:
It's less about what it returns and more about how what is returned gets used. This is the "client" side of the service and we hand this out for other APIs to connect to. Much like port = client port and target port = listen port. It seems like we need that split for address as well.

karolz-ms replied:
Do we need changes or new tests in https://github.com/dotnet/aspire/blob/main/tests/Aspire.Hosting.Tests/Dcp/ApplicationExecutorTests.cs?

karolz-ms replied:
> It's less about what it returns and more about how what is returned gets used. This is the "client" side of the service and we hand this out for other APIs to connect to. Much like port = client port and target port = listen port. It seems like we need that split for address as well.

Not sure I follow. The localhost gets special treatment, but nothing else does. Meaning, if you tell DCP something is listening on address other than localhost (including 0.0.0.0 pseudo-address), it will use that address for the proxy, and this is the address the clients should be using too. So I am not sure why we need to force localhost here.

karolz-ms replied:
Chatted with @davidfowl and understand now, 0.0.0.0, as-is, is not useable by clients. Will need to get that fixed on DCP side.

davidfowl committed:
PR feedback

davidfowl committed:
Revert to localhost for the client address

davidfowl commented:
/azp run

davidfowl committed:
Merge branch 'main' into davidfowl/parse-url
